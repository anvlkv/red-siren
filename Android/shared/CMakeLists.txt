# This goes in the same folder as the build.gradle
project(shared)
cmake_minimum_required(VERSION 3.4.1)
set (SRC_DIR ./src/main/cpp)
add_library (dummy SHARED ${SRC_DIR}/dummy.cpp)
target_link_libraries(dummy)  # Is this needed?

# Note that this all works for apps and for Android libraries (i.e. those with gradle plugin com.android.library).
# However, I don't recommend using this approach for the latter, as a library shouldn't be adding stuff like libc++_shared.so
# itself as this will clash with any other dependency which does the same (two libraries with the same name).