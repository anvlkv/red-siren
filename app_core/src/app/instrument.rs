use std::sync::{Arc, Mutex};

use crux_core::render::Render;
use crux_core::App;
use crux_macros::Effect;
use hecs::{Entity, World};
use serde::{Deserialize, Serialize};

pub use config::Config;
pub use layout::{Layout, LayoutRoot};
use node::spawn_all_nodes;
pub use node::Node;

use crate::{play::Play, Navigate};

pub mod config;
pub mod keyboard;
pub mod layout;
pub mod node;
pub mod string;

#[derive(Default)]
pub struct Instrument;

#[derive(Default)]
pub struct Model {
    pub config: Config,
    pub world: Arc<Mutex<World>>,
    pub inbound: Option<Entity>,
    pub outbound: Option<Entity>,
    pub keyboard: Option<Entity>,
    pub root: Option<Entity>,
    pub nodes: Vec<Entity>,
    pub playing: bool,
    pub layout: Option<Layout>,
    pub setup_complete: bool,
    pub configured: bool,
}

impl Model {
    #[must_use]
    pub fn new(world: Arc<Mutex<World>>) -> Self {
        Self {
            world,
            ..Default::default()
        }
    }
}

#[derive(Default, Serialize, Deserialize, Clone, PartialEq)]
pub struct InstrumentVM {
    pub config: Config,
    pub nodes: Vec<Node>,
    pub playing: bool,
    pub layout: Layout,
}

impl Eq for InstrumentVM {}

#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
pub enum PlaybackEV {
    Play(bool),
    Error,
}

impl Eq for PlaybackEV {}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
pub enum InstrumentEV {
    None,
    CreateWithConfig(Config),
    Playback(PlaybackEV),
    PlayOpPermission(bool),
    PlayOpInstall(bool),
    PlayOpConfigure(bool),
    PlayOpPlay(bool),
    PlayOpPause(bool),
}

#[cfg_attr(feature = "typegen", derive(crux_macros::Export))]
#[derive(Effect)]
#[effect(app = "Instrument")]
pub struct InstrumentCapabilities {
    pub render: Render<InstrumentEV>,
    pub play: Play<InstrumentEV>,
    pub navigate: Navigate<InstrumentEV>,
}

impl App for Instrument {
    type Event = InstrumentEV;

    type Model = Model;

    type ViewModel = InstrumentVM;

    type Capabilities = InstrumentCapabilities;

    fn update(&self, event: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
        match event {
            InstrumentEV::CreateWithConfig(config) => {
                model.config = config.clone();
                let mut world = model.world.lock().expect("world lock");

                let inbound = string::InboundString::spawn(&mut world, &config);
                let outbound = string::OutboundString::spawn(&mut world, &config);
                let keyboard = keyboard::Keyboard::spawn(&mut world, &config);

                let root = layout::LayoutRoot::spawn(&mut world, inbound, outbound, keyboard);

                let layout = Layout::new(&world, &root, &config).expect("Layout failed");
                _ = model.layout.insert(layout);

                _ = model.root.insert(root);
                _ = model.inbound.insert(inbound);
                _ = model.outbound.insert(outbound);
                _ = model.keyboard.insert(keyboard);

                model.nodes = spawn_all_nodes(&mut world);

                if model.setup_complete {
                    let nodes = self.get_nodes(model);
                    caps.play.configure(
                        &model.config,
                        nodes.as_slice(),
                        InstrumentEV::PlayOpConfigure,
                    );
                }

                caps.render.render();
            }
            InstrumentEV::PlayOpPermission(grant) => {
                if grant {
                    caps.play.install_au(InstrumentEV::PlayOpInstall)
                } else {
                    caps.navigate.to(crate::Activity::Intro)
                }
            }
            InstrumentEV::PlayOpInstall(success) => {
                if !success {
                    self.update(InstrumentEV::Playback(PlaybackEV::Error), model, caps)
                } else {
                    model.setup_complete = true;
                    let nodes = self.get_nodes(model);
                    caps.play.configure(
                        &model.config,
                        nodes.as_slice(),
                        InstrumentEV::PlayOpConfigure,
                    );
                }
            }
            InstrumentEV::PlayOpConfigure(success) => {
                model.configured = success;
                if !success {
                    self.update(InstrumentEV::Playback(PlaybackEV::Error), model, caps)
                } else {
                    self.update(
                        InstrumentEV::Playback(PlaybackEV::Play(model.playing)),
                        model,
                        caps,
                    )
                }
            }
            InstrumentEV::PlayOpPause(success) => {
                if !success {
                    self.update(InstrumentEV::Playback(PlaybackEV::Error), model, caps)
                }
            }
            InstrumentEV::PlayOpPlay(success) => {
                if !success {
                    self.update(InstrumentEV::Playback(PlaybackEV::Error), model, caps)
                }
            }
            InstrumentEV::Playback(playback_ev) => match playback_ev {
                PlaybackEV::Play(playing) => {
                    model.playing = playing;
                    if !model.setup_complete {
                        caps.play.permissions(InstrumentEV::PlayOpPermission)
                    } else if !model.configured {
                        let nodes = self.get_nodes(model);
                        caps.play.configure(
                            &model.config,
                            nodes.as_slice(),
                            InstrumentEV::PlayOpConfigure,
                        );
                    } else if playing {
                        caps.play.play(InstrumentEV::PlayOpPlay)
                    } else {
                        caps.play.pause(InstrumentEV::PlayOpPlay)
                    }
                    caps.render.render();
                }
                PlaybackEV::Error => {
                    model.playing = false;
                    model.setup_complete = false;
                    caps.render.render();
                }
            },
            InstrumentEV::None => {}
        }
    }

    fn view(&self, model: &Self::Model) -> Self::ViewModel {
        InstrumentVM {
            nodes: self.get_nodes(model),
            playing: model.playing,
            config: model.config.clone(),
            layout: model.layout.clone().unwrap_or_default(),
        }
    }
}

impl Instrument {
    fn get_nodes(&self, model: &Model) -> Vec<Node> {
        let world = model.world.lock().expect("world lock");
        model
            .nodes
            .iter()
            .map(|e| *world.get::<&Node>(*e).expect("node for entity"))
            .collect()
    }
}
