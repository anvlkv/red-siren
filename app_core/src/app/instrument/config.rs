use std::collections::BTreeMap;
use hecs::{Entity, World, Bundle};
use serde::{Deserialize, Serialize};

const MIN_BUTTON_SIZE_IN: f64 = 0.75;
const MAX_BUTTON_SIZE_B_RATIO: f64 = 0.6;
const BUTTON_TRACK_MARGIN_RATION: f64 = 0.2;
const BUTTON_SPACE_RATIO: f64 = 2.0;
const DPI_RANGE: &[usize] = &[120, 160, 240, 320, 480, 640];
const F_BASE: f64 = 110.0;
const F_MAX: f64 = 5500.0;

#[derive(Debug, Default, Serialize, Deserialize, Clone, PartialEq, Bundle)]
pub struct Config {
    pub portrait: bool,
    pub width: f64,
    pub height: f64,
    pub breadth: f64,
    pub length: f64,
    pub whitespace: f64,
    pub groups: usize,
    pub buttons_group: usize,
    pub n_buttons: usize,
    pub button_size: f64,
    pub button_track_margin: f64,
    pub safe_area: [f64; 4],
    pub f0: f32,
}

impl Eq for Config {}

impl Config {
    pub fn new(width: f64, height: f64, dpi: f64, safe_area: [f64; 4]) -> Self {
        let portrait = height > width;

        let (length, safe_length, safe_breadth) = if portrait {
            (
                height,
                height - safe_area[1] - safe_area[3],
                (width / 3.0).min(width - safe_area[0] - safe_area[2]),
            )
        } else {
            (
                width,
                width - safe_area[0] - safe_area[2],
                (height / 3.0).min(height - safe_area[1] - safe_area[3]),
            )
        };

        let max_button_size = (safe_breadth * MAX_BUTTON_SIZE_B_RATIO).round() as usize;
        let min_button_size = f64::sqrt(dpi * MIN_BUTTON_SIZE_IN).round() as usize;

        let (min_groups, min_buttons) = {
            let max_buttons = ((length - max_button_size as f64 * BUTTON_SPACE_RATIO)
                / max_button_size as f64)
                .round() as usize;
            let slots = max_buttons.div_euclid(2);
            vec![
                (slots.div_euclid(5), 5),
                (slots.div_euclid(3), 3),
                (slots.div_euclid(2), 2),
            ]
            .into_iter()
            .fold((1, 1), |acc, (groups, buttons_group)| {
                if groups * buttons_group > acc.0 * acc.1
                    || groups * buttons_group == acc.0 * acc.1 && buttons_group > acc.1
                {
                    (groups, buttons_group)
                } else {
                    acc
                }
            })
        };

        let min_count = min_groups * min_buttons;

        let f_c = (F_BASE / f64::sqrt((length * safe_breadth) / dpi)) as f32;

        let f0 = {
            let mut f0 = f_c;

            while f0 < F_BASE as f32 {
                f0 *= 2.0;
            }

            f0
        };

        let mut candidates = Vec::<(usize, usize, usize, usize)>::new();

        for size in min_button_size..=max_button_size {
            let space = size as f64 * BUTTON_SPACE_RATIO * 2.0;
            let active_length = (safe_length - space).round() as usize;
            let slots = num_integer::gcd(space.round() as usize + size, active_length);
            for (groups, buttons_group) in [
                (slots.div_euclid(5), 5),
                (slots.div_euclid(3), 3),
                (slots.div_euclid(2), 2),
            ] {
                let count = groups * buttons_group;
                let used_space = space * count as f64;
                let f_max = f0 as f64 * 2.0 * (groups * buttons_group) as f64;
                if count >= min_count && used_space < safe_length && f_max <= F_MAX {
                    candidates.push((size, groups, buttons_group, active_length));
                }
            }
        }

        let (d_size, d_groups, d_buttons_group, d_active_length, d_count) = candidates.iter().fold(
            ((0, 0), (0, 0), (0, 0), (0, 0), (0, 0)),
            |mut acc, (size, groups, buttons_group, active_length)| {
                let count = groups * buttons_group;
                acc.0 = (acc.0 .0.min(*size), acc.0 .1.max(*size));
                acc.1 = (acc.1 .0.min(*groups), acc.1 .1.max(*groups));
                acc.2 = (acc.2 .0.min(*buttons_group), acc.2 .1.max(*buttons_group));
                acc.3 = (acc.3 .0.min(*active_length), acc.3 .1.max(*active_length));
                acc.4 = (acc.4 .0.min(count), acc.4 .1.max(count));
                acc
            },
        );

        let (d_size, d_groups, d_buttons_group, d_active_length, d_count) = (
            (d_size.1 - d_size.0) as f64,
            (d_groups.1 - d_groups.0) as f64,
            (d_buttons_group.1 - d_buttons_group.0) as f64,
            (d_active_length.1 - d_active_length.0) as f64,
            (d_count.1 - d_count.0).max(1) as f64,
        );

        let candidates =
            BTreeMap::<usize, (usize, usize, usize, usize)>::from_iter(candidates.into_iter().map(
                |(size, groups, buttons_group, active_length)| {
                    let count = (buttons_group * groups) as f64;
                    let score = (size as f64 / d_size
                        + groups as f64 / d_groups
                        + buttons_group as f64 / d_buttons_group
                        + active_length as f64 / d_active_length)
                        * (1.0 + d_count / count);

                    (
                        score.round() as usize,
                        (size, groups, buttons_group, active_length),
                    )
                },
            ));

        let scores = candidates.keys().collect::<Vec<_>>();

        log::debug!("scores: {scores:?}");

        let dpi_pos = DPI_RANGE
            .iter()
            .position(|d| *d as f64 >= dpi)
            .map_or(0.0, |pos| {
                candidates.len() as f64
                    - ((pos + 1) as f64 / DPI_RANGE.len() as f64) * candidates.len() as f64
            })
            .abs()
            .round() as usize;

        log::debug!("dpi_pos: {dpi_pos}");

        let (button_size, groups, buttons_group, active_length) =
            candidates.into_iter().nth(dpi_pos).map_or(
                (max_button_size as f64, min_groups, min_buttons, safe_length),
                |(_, (button_size, groups, buttons_group, active_length))| {
                    (
                        button_size as f64,
                        groups,
                        buttons_group,
                        active_length as f64,
                    )
                },
            );

        let whitespace = (safe_length - active_length) / 2.0;

        Config {
            portrait,
            width,
            height,
            length: active_length,
            breadth: safe_breadth,
            button_size,
            groups,
            buttons_group,
            n_buttons: buttons_group * groups,
            button_track_margin: BUTTON_TRACK_MARGIN_RATION,
            safe_area,
            whitespace,
            f0,
        }
    }

    pub fn spawn(&self, world: &mut World) -> Entity {
        world.spawn((self.clone(),))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const RAND_SCREENS: &[(f64, f64, f64)] = &[
        (1920.0, 1080.0, 96.0),
        (2560.0, 1440.0, 110.0),
        (1366.0, 768.0, 125.0),
        (2732.0, 2048.0, 264.0),
        (2436.0, 1125.0, 458.0),
        (2960.0, 1440.0, 568.0),
        (3840.0, 2160.0, 163.0),
        (1280.0, 800.0, 180.0),
        (3440.0, 1440.0, 110.0),
        (2560.0, 1600.0, 227.0),
        (1080.0, 2340.0, 394.0),
        (6016.0, 3384.0, 220.0),
        (2048.0, 1536.0, 264.0),
        (2960.0, 1440.0, 522.0),
        (1280.0, 720.0, 267.0),
        (2560.0, 1440.0, 163.0),
        (1280.0, 1024.0, 96.0),
        (3840.0, 1080.0, 110.0),
        (2224.0, 1668.0, 264.0),
        (2960.0, 1440.0, 570.0),
        (3840.0, 1600.0, 163.0),
        (1280.0, 720.0, 326.0),
        (1920.0, 1200.0, 224.0),
        (2560.0, 1440.0, 141.0),
        (1366.0, 768.0, 100.0),
        (2560.0, 1440.0, 440.0),
        (1280.0, 800.0, 149.0),
        (2960.0, 1440.0, 522.0),
        (3840.0, 2160.0, 204.0),
        (2560.0, 1600.0, 197.0),
    ];

    #[test]
    fn config_snapshot_by_rand_screen() {
        let mut configs = Vec::new();
        for config in RAND_SCREENS.iter().map(|(width, height, dpi)| {
            Config::new(*width, *height, *dpi, [50.0, 20.0, 10.0, 25.0])
        }) {
            configs.push(config);
        }

        insta::assert_yaml_snapshot!(configs)
    }
}
