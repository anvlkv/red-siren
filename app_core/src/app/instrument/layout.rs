use anyhow::Result;
use hecs::{Bundle, Entity, World};
use keyframe::CanTween;
use serde::{Deserialize, Serialize};

use crate::geometry::{line::Line, rect::Rect};

use super::{
    keyboard::{Button, ButtonGroup, Keyboard, Track},
    string::{InboundString, OutboundString},
    Config,
};

#[derive(Default, Serialize, Deserialize, Clone, PartialEq, Debug, Eq)]
pub struct Layout {
    pub inbound: Line,
    pub outbound: Line,
    pub buttons: Vec<Rect>,
    pub tracks: Vec<Rect>,
    pub menu_position: MenuPosition,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug, Eq)]
pub enum MenuPosition {
    TopLeft(Rect),
    TopRight(Rect),
    BottomLeft(Rect),
    Center(Rect),
}

impl Default for MenuPosition {
    fn default() -> Self {
        Self::Center(Rect::default())
    }
}

impl MenuPosition {
    pub fn rect(&self) -> &Rect {
        match self {
            MenuPosition::Center(r)
            | MenuPosition::TopLeft(r)
            | MenuPosition::TopRight(r)
            | MenuPosition::BottomLeft(r) => r,
        }
    }
}

impl CanTween for MenuPosition {
    fn ease(from: Self, to: Self, time: impl keyframe::num_traits::Float) -> Self {
        let r1 = *from.rect();

        match to {
            MenuPosition::TopLeft(r) => MenuPosition::TopLeft(CanTween::ease(r1, r, time)),
            MenuPosition::Center(r) => MenuPosition::Center(CanTween::ease(r1, r, time)),
            MenuPosition::TopRight(r) => MenuPosition::TopRight(CanTween::ease(r1, r, time)),
            MenuPosition::BottomLeft(r) => MenuPosition::BottomLeft(CanTween::ease(r1, r, time)),
        }
    }
}

#[derive(Bundle)]
pub struct LayoutRoot {
    entities: (Entity, Entity, Entity),
}

impl LayoutRoot {
    pub fn spawn(world: &mut World, inbound: Entity, outbound: Entity, keyboard: Entity) -> Entity {
        world.spawn((Self {
            entities: (inbound, keyboard, outbound),
        },))
    }

    pub fn inbound(&self) -> Entity {
        self.entities.0
    }
    pub fn outbound(&self) -> Entity {
        self.entities.2
    }
    pub fn keyboard(&self) -> Entity {
        self.entities.1
    }
}

impl CanTween for Layout {
    fn ease(from: Self, to: Self, time: impl keyframe::num_traits::Float) -> Self {
        let outbound = CanTween::ease(from.outbound, to.outbound, time);
        let inbound = CanTween::ease(from.inbound, to.inbound, time);

        let buttons = ease_vec(from.buttons, to.buttons, time);
        let tracks = ease_vec(from.tracks, to.tracks, time);
        let menu_position = CanTween::ease(from.menu_position, to.menu_position, time);

        Self {
            outbound,
            inbound,
            buttons,
            tracks,
            menu_position,
        }
    }
}

impl Layout {
    pub fn new(world: &World, root: &Entity, config: &Config) -> Result<Self> {
        let root = world.get::<&LayoutRoot>(*root)?;
        let (inbound, keyboard, outbound) = (
            world.get::<&InboundString>(root.inbound())?,
            world.get::<&Keyboard>(root.keyboard())?,
            world.get::<&OutboundString>(root.outbound())?,
        );

        let (buttons, tracks): (Vec<Rect>, Vec<Rect>) = keyboard
            .groups
            .iter()
            .enumerate()
            .try_fold(
                Vec::<(Rect, Rect)>::new(),
                |mut acc, (_i, g)| -> Result<_> {
                    let group = world.get::<&ButtonGroup>(*g)?;
                    for b in &group.buttons {
                        let button = world.get::<&Button>(*b)?;
                        let track = world.get::<&Track>(button.track)?;
                        acc.push((button.rect, track.rect));
                    }
                    Ok(acc)
                },
            )?
            .into_iter()
            .unzip();

        log::debug!("tracks {tracks:#?}");

        let menu_position = tracks
            .first()
            .map(|t| {
                if config.portrait {
                    if t.top_left().x >= config.breadth {
                        MenuPosition::TopLeft(Rect::new(
                            config.safe_area[0],
                            config.breadth,
                            config.safe_area[1],
                            config.breadth,
                        ))
                    } else {
                        MenuPosition::TopRight(Rect::new(
                            config.width - config.breadth,
                            config.width - config.safe_area[2],
                            config.safe_area[1],
                            config.breadth,
                        ))
                    }
                } else if t.top_left().y < config.breadth {
                    MenuPosition::BottomLeft(Rect::new(
                        config.safe_area[0],
                        config.breadth,
                        config.height - config.breadth,
                        config.height - config.safe_area[3],
                    ))
                } else {
                    MenuPosition::TopLeft(Rect::new(
                        config.safe_area[0],
                        config.breadth,
                        config.safe_area[1],
                        config.breadth,
                    ))
                }
            })
            .unwrap_or_default();

        Ok(Self {
            inbound: inbound.line,
            outbound: outbound.line,
            buttons,
            tracks,
            menu_position,
        })
    }

    pub fn dummy(breadth: f64, length: f64, button_size: f64) -> Self {
        let inbound = Line::new(0.0, length, 0.0, 1.0);
        let outbound = Line::new(0.0, length, breadth, breadth + 1.0);

        let button = Rect::new(0.0, button_size, 0.0, button_size);

        Self {
            inbound,
            outbound,
            buttons: vec![button],
            ..Default::default()
        }
    }
}

fn ease_vec(from: Vec<Rect>, to: Vec<Rect>, time: impl keyframe::num_traits::Float) -> Vec<Rect> {
    let len = CanTween::ease(from.len() as f64, to.len() as f64, time).round() as usize;
    (0..len)
        .map(|i| {
            let (from_b, to_b) = (from.get(i), to.get(i));
            match (from_b, to_b) {
                (Some(from), Some(to)) => CanTween::ease(*from, *to, time),
                (Some(from), None) => {
                    let c = from.center();
                    let z_rect = Rect::new(c.x, c.x, c.y, c.y);
                    CanTween::ease(*from, z_rect, time)
                }
                (None, Some(to)) => {
                    let c = to.center();
                    let z_rect = Rect::new(c.x, c.x, c.y, c.y);
                    CanTween::ease(z_rect, *to, time)
                }
                (None, None) => unreachable!(),
            }
        })
        .collect()
}
