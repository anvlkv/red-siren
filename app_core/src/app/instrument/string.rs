use hecs::{Bundle, Entity, World};
use mint::Point2;

use crate::geometry::Line;

use super::Config;

#[derive(Default, Bundle)]
pub struct InboundString {
    pub line: Line,
}

impl InboundString {
    pub fn spawn(world: &mut World, config: &Config) -> Entity {
        world.spawn((InboundString {
            line: string_line(config, 2.0),
        },))
    }
}

#[derive(Default, Bundle)]
pub struct OutboundString {
    pub line: Line,
    pub data: Vec<Point2<f64>>,
}

impl OutboundString {
    pub fn spawn(world: &mut World, config: &Config) -> Entity {
        world.spawn((OutboundString {
            line: string_line(config, 1.0),
            data: vec![],
        },))
    }

    pub fn update_data(&mut self, data: Vec<f32>, config: &Config) {
        let l_step = config.length / data.len() as f64;
        let b_step = config.breadth / 2.25;
        let b_base = if config.portrait {
            self.line.p0().x
        } else {
            self.line.p0().y
        };

        self.data = data.into_iter().enumerate().map(|(i, val)| {
            let val = val * 512.0;
            if config.portrait {
                Point2 {
                    x: b_base + b_step * val as f64,
                    y: i as f64 * l_step,
                }
            } else {
                Point2 {
                    y: b_base + b_step * val as f64,
                    x: i as f64 * l_step,
                }
            }
        }).collect();
    }
}

fn string_line(config: &Config, at: f64) -> Line {
    if config.portrait {
        let main = (config.width - config.breadth) / at;
        Line::new(
            main,
            main,
            0.0,
            config.length + config.safe_area[3] + config.safe_area[1] + config.whitespace * 2.0,
        )
    } else {
        let main = (config.height - config.breadth) / at;
        Line::new(
            0.0,
            config.length + config.safe_area[2] + config.safe_area[0] + config.whitespace * 2.0,
            main,
            main,
        )
    }
}
