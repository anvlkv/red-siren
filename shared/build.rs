fn main() {
    #[cfg(not(feature = "cargo-clippy"))]
    uniffi::generate_scaffolding("./src/shared.udl").unwrap();
}
