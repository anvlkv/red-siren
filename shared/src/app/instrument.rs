use serde::{Deserialize, Serialize};

use crux_core::render::Render;
use crux_core::App;
use crux_kv::KeyValue;
use crux_macros::Effect;

#[derive(Default)]
pub struct Instrument;

#[derive(Default, Serialize, Deserialize)]
pub struct Model {
    pub config: Config,
}

#[derive(Debug, Default, Serialize, Deserialize, Clone, Hash, PartialEq, Eq)]
pub struct Config {
    pub portrait: bool,
    pub width: u16,
    pub height: u16,
    pub groups: u8,
    pub buttons_group: u8,
    pub button_size: u8,
    pub gap: (u8, u8),
    pub steps: u16,
}

impl Config {
    pub fn new(
        width: usize,
        height: usize,
        button_size: usize,
        rigid_step: bool,
        max_buttons: usize,
    ) -> Self {
        let portrait = height > width;
        let (length, breadth) = if portrait {
            (height, width)
        } else {
            (width, height)
        };

        let mut button_gap = button_size / 2;
        let mut button_group_gap = button_gap + button_size;

        let steps = if rigid_step {
            breadth / 2 / button_size
        } else {
            breadth / 2
        };

        let active_length = length - button_group_gap * 2;

        let groups: Vec<(Vec<(usize, usize)>, usize)> = (1..=active_length)
            .into_iter()
            .fold(Vec::<(Vec<(usize, usize)>, usize)>::new(), |mut acc, _| {
                if acc.last().is_some() {
                    let pos = acc.len() - 1;
                    let group = &mut acc[pos];
                    let group_len = group.0.len();

                    let mut last = group.0.last_mut();
                    let btn = last.as_mut().unwrap();
                    if btn.0 < button_size {
                        btn.0 += 1;
                    } else if btn.1 < button_gap && group_len < max_buttons {
                        btn.1 += 1;
                    } else if group_len < max_buttons {
                        group.0.push((1, 0));
                    } else if group.1 < button_group_gap {
                        group.1 += 1;
                    } else {
                        acc.push((vec![(1, 0)], 0));
                    }
                } else {
                    acc.push((vec![(1, 0)], 0));
                }
                acc
            })
            .into_iter()
            .map(|group| {
                (
                    group
                        .0
                        .into_iter()
                        .filter(|btn| btn.0 == button_size)
                        .collect(),
                    group.1,
                )
            })
            .collect();

        let optimal_buttons = groups.iter().map(|g| g.0.len()).max().unwrap_or(1);
        let optimal_groups = groups
            .iter()
            .filter(|g| g.0.len() == optimal_buttons)
            .count();

        if optimal_groups < groups.len() {
            let empty = groups.iter().skip(optimal_groups).fold(0, |acc, group| acc + group.1 + group.0.iter().fold(0, |a, b| a+b.0+b.1));
            let extra = empty / (optimal_groups * optimal_buttons);

            button_gap = button_gap + extra / optimal_groups;
            button_group_gap = button_gap + button_size;
        }

        Config {
            button_size: button_size.try_into().unwrap_or(42),
            portrait,
            width: width.try_into().unwrap_or(200),
            height: height.try_into().unwrap_or(200),
            steps: steps.try_into().unwrap_or(2),
            gap: (
                button_gap.try_into().unwrap_or(24),
                button_group_gap.try_into().unwrap_or(66),
            ),
            groups: optimal_groups.try_into().unwrap_or(1),
            buttons_group: optimal_buttons.try_into().unwrap_or(1),
        }
    }
}

#[derive(Default, Serialize, Deserialize, Clone, Hash, PartialEq, Eq)]
pub struct ViewModel {
    pub config: Config,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
pub enum Event {
    None,
}

#[cfg_attr(feature = "typegen", derive(crux_macros::Export))]
#[derive(Effect)]
#[effect(app = "Instrument")]
pub struct InstrumentCapabilities {
    pub render: Render<Event>,
    pub key_value: KeyValue<Event>,
}

impl App for Instrument {
    type Event = Event;

    type Model = Model;

    type ViewModel = ViewModel;

    type Capabilities = InstrumentCapabilities;

    fn update(&self, event: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
        todo!()
    }

    fn view(&self, model: &Self::Model) -> Self::ViewModel {
        ViewModel::default()
    }
}
