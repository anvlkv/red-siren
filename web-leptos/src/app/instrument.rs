mod button;
mod string;
mod track;
mod menu;
mod playback;

use leptos::*;
use shared::{instrument, key_value::KeyValueOutput};

pub use button::ButtonComponent;
pub use menu::MenuComponent;
pub use string::StringComponent;
pub use track::TrackComponent;

use crate::{app::instrument::playback::PlayBackState, kv::KVContext};

#[component]
pub fn InstrumentComponent(
    vm: Signal<instrument::InstrumentVM>,
    ev: SignalSetter<instrument::InstrumentEV>,
) -> impl IntoView {
    let view_box = move || {
        let vb = vm().view_box;
        format!(
            "{} {} {} {}",
            vb.top_left().x,
            vb.top_left().y,
            vb.bottom_right().x,
            vb.bottom_right().y
        )
    };

    let inbound_layout_line = Signal::derive(move || vm().layout.inbound);
    let outbound_layout_line = Signal::derive(move || vm().layout.outbound);

    let playing = Signal::derive(move || vm().playing);
    let config = Signal::derive(move || vm().config);
    let playback_ev = SignalSetter::map(move |e| ev.set(instrument::InstrumentEV::Playback(e)));
    let playback_state = playback::create_playback(playing.clone(), config, playback_ev);

    // create_effect(move |prev| {
    //     let state = playback_state.get();

    //     match (state, prev) {
    //         (PlayBackState::Playing, Some(PlayBackState::Preparing)) => ev(instrument::InstrumentEV::Input(
    //             KeyValueOutput::Read(kv_ref.borrow_mut().remove(instrument::INPUT_STREAM_KV)),
    //         )),
    //         _ => {}
    //     }

    //     state
    // });

    let toggle_playing = Callback::<()>::new(move |_| {
        ev(instrument::InstrumentEV::Playback(
            instrument::PlaybackEV::Play(!playing()),
        ))
    });
    let menu_position = Signal::derive(move || vm().layout.menu_position);

    view! {
      <div class="h-full w-full bg-red dark:bg-black instrument">
        <svg fill="none" class="stroke-black dark:stroke-red" viewBox={view_box} xmlns="http://www.w3.org/2000/svg">
          <StringComponent layout_line={inbound_layout_line} />
          <StringComponent layout_line={outbound_layout_line} />
        </svg>
        <svg class="fill-red dark:fill-black stroke-black dark:stroke-red" viewBox={view_box} xmlns="http://www.w3.org/2000/svg">
          {move || vm().layout.tracks.into_iter().zip(vm().nodes).map(|(rect, node)|
            view!{
              <TrackComponent layout_rect={Signal::derive(move || rect)}/>
            }
          ).collect_view()}
        </svg>
        <svg class="fill-black dark:fill-red" viewBox={view_box} xmlns="http://www.w3.org/2000/svg">
          {move || vm().layout.buttons.into_iter().zip(vm().nodes).map(|(rect, node)|
            view!{
              <ButtonComponent layout_rect={Signal::derive(move || rect)} f_n={node.f_n}/>
            }
          ).collect_view()}
        </svg>
        <MenuComponent position={menu_position} playing=playing toggle_playing=toggle_playing />
      </div>
    }
}
