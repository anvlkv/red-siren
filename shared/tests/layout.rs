use hecs::World;
use shared::instrument::{keyboard, string, Config, Layout, LayoutRoot};

#[test]
fn creates_layout_from_config() {
    let config = Config::new(430.0, 932.0, 476.0, Default::default());
    let mut world = World::new();

    let inbound = string::InboundString::spawn(&mut world, &config);
    let outbound = string::OutboundString::spawn(&mut world, &config);
    let keyboard = keyboard::Keyboard::spawn(&mut world, &config);

    let root = LayoutRoot::spawn(&mut world, inbound, outbound, keyboard);
    let layout = Layout::new(&world, &root).expect("failed to create layout");

    insta::assert_yaml_snapshot!(layout)
}
