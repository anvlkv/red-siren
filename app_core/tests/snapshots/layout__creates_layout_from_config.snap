---
source: shared/tests/layout.rs
expression: layout
---
inbound:
  line:
    - - 143.33333333333331
      - 0
    - - 143.33333333333331
      - 932
outbound:
  line:
    - - 286.66666666666663
      - 0
    - - 286.66666666666663
      - 932
buttons:
  - rect:
      - - 193.5
        - 140.5
      - - 236.5
        - 183.5
  - rect:
      - - 193.5
        - 292.5
      - - 236.5
        - 335.5
  - rect:
      - - 193.5
        - 444.5
      - - 236.5
        - 487.5
  - rect:
      - - 193.5
        - 596.5
      - - 236.5
        - 639.5
  - rect:
      - - 193.5
        - 748.5
      - - 236.5
        - 791.5
tracks:
  - rect:
      - - 184.9
        - 131.9
      - - 574.7666666666667
        - 192.1
  - rect:
      - - 184.9
        - 283.9
      - - 574.7666666666667
        - 344.1
  - rect:
      - - 184.9
        - 435.9
      - - 574.7666666666667
        - 496.1
  - rect:
      - - 184.9
        - 587.9
      - - 574.7666666666667
        - 648.1
  - rect:
      - - 184.9
        - 739.9
      - - 574.7666666666667
        - 800.1
menu_position:
  TopLeft:
    rect:
      - - 0
        - 0
      - - 143.33333333333334
        - 143.33333333333334

